### PR TITLE
Feature/enhance multiline

### PR DESCRIPTION
**Pull request by gpwclark**

### CODE REVIEW

**Comment by gpwclark** [comment]
this is ready to go now.

File: examples/dodo_prompt.rs [hunk 1]
```diff
@@ -77,8 +82,73 @@ impl Completer for CommentCompleter {
...
+
+impl NewlineRule for NewlineForBackslashAndOpenDelimRule {
+    fn evaluate_on_newline(&self, buf: &Buffer) -> bool {
+        last_non_ws_char_was_not_backslash(buf) && check_balanced_delimiters(buf)
+    }
+}
+
+pub fn check_balanced_delimiters(buf: &Buffer) -> bool {
+    let buf_vec = buf.range_graphemes_all();
+    let mut stack = vec![];
```
Comment (sstanfield) [inline examples/dodo_prompt.rs:95]: Can't this just be a counter instead of a vec, it looks like there is no need to allocate a vector for this based on it's use.

File: examples/dodo_prompt.rs [hunk 2]
```diff
@@ -77,8 +82,73 @@ impl Completer for CommentCompleter {
...
+    let mut symmetric_delimiters = HashSet::new();
```
Comment (sstanfield) [inline examples/dodo_prompt.rs:96]: As above, this also looks like it could just be a counter.

File: src/buffer.rs [hunk 3]
```diff
@@ -789,7 +779,7 @@ mod tests {
     fn test_buffer_to_string_ignore_newline() {
         let mut buf = Buffer::new();
         buf.insert(0, ['h', 'e', '\\', '\n', 'l', 'l', 'o'].iter());
-        assert_eq!("hello".to_owned(), String::from(buf));
+        assert_eq!("he\\\nllo".to_owned(), String::from(buf));
```
Comment (sstanfield) [inline src/buffer.rs:782]: Why did this test change?  Would not a backslash at the end of a line still "consume" the newline?

File: src/editor.rs [hunk 4]
```diff
@@ -6,16 +6,15 @@ use crate::context::ColorClosure;
...
 use crate::{Buffer, Cursor};
 use crate::{History, Metrics};
 
-use super::complete::Completer;
-
 /// The core line editor. Displays and provides editing for history and the new buffer.
 pub struct Editor<'a> {
     prompt: Prompt,
     history: &'a mut History,
+    editor_rules: &'a dyn EditorRules,
```
Comment (sstanfield) [inline src/editor.rs:17]: You could use a generic here and avoid the lifetime.  I.e. pub struct Editor<T: EditorRules> ... editor_rules: T and hand ownership to Editor.  Not strictly needed but make it a little easier to use Editor if you don't need worry about the lifetime tied to an EditorRules instance.

Comment (gpwclark) [reply src/editor.rs:17]: ok, so, initially I avoided this because introducing a type parameter to editor was a silly amount of signature changes. and my main goal was to make sure the DefaultEditorRule was using generics such that it was easier for the client to move their own implementations in:
```rust 
    let mut con = Context::new();
    let editor_rules = DefaultEditorRules::custom(
        DefaultWordDivideRule {},
        NewlineForBackslashAndOpenDelimRule {},
    ); // this is the part I used generics for, and I like that the client just move their implementations in
    con.set_editor_rules(Box::new(editor_rules)); // but we're still using a trait object type here
```

Today I bit the bullet and started changing everything in Editor to be generic and ended up in a weird corner. I put the code in  a separate branch because we eventually might want this and it's very tedious to do: 

https://github.com/sl-sh-dev/sl-liner/tree/feature/enhance-multiline-genericize

the compilation is failing because Completer and KeyMap are not type object safe when they have methods that use generics. I think I understand why: 

https://doc.rust-lang.org/reference/items/traits.html#object-safety

but to actually make Editor use a generic EditorRules, assuming KeyMap and Completer are going to need to refer to the Editor, which is 100% going to be true going forward, then KeyMap and Completer are also going to have to be made generic (right?), and that might be out of scope for this PR. And I'm not sure the sheer amount of code changes justify the win of not having a lifetime.

File: src/buffer.rs (hunk 3 above)
Comment (gpwclark) [reply src/buffer.rs]: I removed this test. It doesn't make sense anymore. It's just testing this method

```

impl From<Buffer> for String {
    fn from(buf: Buffer) -> Self {
        buf.data
    }
}
```

and that method used to strip newlines.

We decided this was bad because it meant that autocompletion did not include newlines, which was confusing.  As it stands now that test technically showed us the old code doesn't exist. 

The other thing that's confusing is the buffer doesn't, and I don't know if it should or shouldn't, understand anything about multiple lines. It depends on future development. Going forward, when doing multi line editing we want up and down to behave the way you'd expect in the terminal and currently they do not. But I don't know if we are going to make buffer multiline aware, and have the keymap implementations "deal" with that. or keep buffer stupid, where it just holds a string, and that string might happen to have some newlines in it, that the keymap will have to work around. or maybe there will be another layer in-between the two, that will take care of that. I don't know yet.

File: examples/dodo_prompt.rs (hunk 1 above)
Comment (gpwclark) [reply examples/dodo_prompt.rs]: uh, this code was not even correct. I spruced it up to be correct, and is the kind of thing that we could now use in sl-sh if we want. The old code didn't actually care if a ( was matched with another ), you could have matched it with a } or a ] and it would have worked. I also added the concept of being inside of a double quote so we were not trying to balance delimiters in user strings.

The thing I decided to not do, was refuse to evalute if we had this business

```
$> cargo run --example dodo_prompt history
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/examples/dodo_prompt history`
History file: history
[prompt]
% ls (
  ]]]]]]]]
  )
ENTERED: [ls (
]]]]]]]]
)]
[prompt]
%
```

basically if YOU open a delimiter, we'll make sure it has a matching close, but if you add closing delimiters that don't have corresponding opening ones... whatever.

My thought process is that it's just less intuitive. if you think something should have evaluates when you hit enter and it did not then the first assumption is that something was opened but not closed. not that something was closed that was never opened. 

Now that I think about it, maybe this idea of `strict balance` becomes optional? TBH I don't know if I'd want to enable it, but it does have some value.

File: examples/dodo_prompt.rs (hunk 2 above)
Comment (gpwclark) [reply examples/dodo_prompt.rs]: see other response.
